src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+
+            EndStream =     0b00000001,
+            Ack =           0b00000001,
+            EndHeaders =    0b00000100,
+            Padded =        0b00001000,
+            Priority =      0b00100000,
+        }
+
+        private async Task SendFramesAsync(Memory<byte> frame)
+        {
+            // This can be called simultaneously by multiple Http2Streams sending their headers/request body,
+            // or by the connection itself to send connection level frames like SETTINGS.
+            // Serialize the writes.
+            await _writerLock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                await _stream.WriteAsync(frame).ConfigureAwait(false);
+            }
+            finally
+            {
+                _writerLock.Release();
+            }
+        }
+
+        // Note that this is safe to be called concurrently by multiple threads.
+
+        public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, bool doRequestAuth, CancellationToken cancellationToken)
+        {
+            int streamId = GetNewStreamId();
+            if (streamId == -1)
+            {
+                // TODO: Gracefully shutdown connection and mark as not usable.
+                // TODO: In debug builds, we should set the stream limit reasonably low (4K maybe?) so that we can hit it during stress testing.
+                throw new Exception("out of stream IDs");
+            }
+
+            Http2Stream http2Stream = new Http2Stream(this, streamId);
+            try
+            {
+                return await http2Stream.SendAsync(request, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                http2Stream.Dispose();
+                throw;
+            }
+        }
+
+        class Http2Stream : IDisposable
+        {
+            private readonly Http2Connection _connection;
+            private readonly int _streamId;
+
+            // TODO: In debug build, make initial size small (10)
+            private ArrayBuffer _requestBuffer;
+
+            private ArrayBuffer _responseBuffer;
+            private object _responseBufferLock;
+            private TaskCompletionSource<bool> _responseDataAvailable;
+            private bool _responseComplete;
+
+            private HttpResponseMessage _response;
+            private bool _disposed = false;
+
+
+            public Http2Stream(Http2Connection connection, int streamId)
+            {
+                _connection = connection;
+                _streamId = streamId;
+
+                if (!_connection._httpStreams.TryAdd(streamId, this))
+                {
+                    Debug.Fail("_httpStreams.TryAdd failed");
+                    throw new InternalException();
+                }
+
+                _requestBuffer = new ArrayBuffer(InitialBufferSize);
+
+                _responseBuffer = new ArrayBuffer(InitialBufferSize);
```

Comment:
We allocate a new Http2Stream for every request made on a connection?  If so, this seems like a lot of allocation, and we really need to pool these so every request doesn't end up incurring an additional 8K of buffers allocated (plus the Http2Stream itself... maybe we can pool those instead?)

Reply:
Yes, I think we need to pool the Http2Stream objects. I'll add a TODO for this.